Cannot serialize _CachedRelation

### Reproduction steps

Run `test__postgres__simple_copy_with_materialized_views` with JSON-formatted logging:
```
DBT_LOG_FORMAT=json python3 -m pytest -m profile_postgres test/integration/001_simple_copy_test/test_simple_copy.py::TestSimpleCopy::test__postgres__simple_copy_with_materialized_views 
```
```python
core/dbt/events/functions.py:208: in create_log_line
    return create_json_log_line(e, msg_fn)  # json output, both console and file
core/dbt/events/functions.py:200: in create_json_log_line
    values = event_to_serializable_dict(e, lambda _: e.get_ts_rfc3339(), lambda x: msg_fn(x))
core/dbt/events/functions.py:139: in event_to_serializable_dict
    for field, value in dataclasses.asdict(e).items():  # type: ignore[attr-defined]
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/dataclasses.py:1073: in asdict
    return _asdict_inner(obj, dict_factory)
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/dataclasses.py:1080: in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/dataclasses.py:1114: in _asdict_inner
    return copy.deepcopy(obj)
/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/copy.py:153: in deepcopy
    y = copier(memo)
core/dbt/adapters/cache.py:82: in __deepcopy__
    new = self.__class__(self.inner.incorporate())
core/dbt/contracts/relation.py:51: in incorporate
    value = self.to_dict(omit_none=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <PostgresRelation "dbt"."test16384599476280903696_simple_copy_001"."unrelated_materialized_view">, use_bytes = False, use_enum = False
use_datetime = False

>   ???
E   AttributeError: 'str' object has no attribute '_serialize'

<string>:14: AttributeError
```

In particular, the `AddRelation` event (included because `--log-cache-events` is enabled) has a `_CachedRelation` object as a dataclass property. This seems like the same issue we were observing in https://github.com/dbt-labs/dbt-core/pull/4361#discussion_r759353979:
```python
ipdb> dataclasses.asdict(e)
*** AttributeError: 'str' object has no attribute '_serialize'
ipdb> e.__dataclass_fields__
{'relation': Field(name='relation',type=<class 'dbt.events.stubs._CachedRelation'>,default=<dataclasses._MISSING_TYPE object at 0x10a541640>,default_factory=<dataclasses._MISSING_TYPE object at 0x10a541640>,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD), 'code': Field(name='code',type=<class 'str'>,default='E024',default_factory=<dataclasses._MISSING_TYPE object at 0x10a541640>,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),_field_type=_FIELD)}
```

--- a/core/dbt/adapters/base/impl.py
+++ b/core/dbt/adapters/base/impl.py
@@ -291,7 +291,7 @@ class BaseAdapter(metaclass=AdapterMeta):
         if (database, schema) not in self.cache:
             fire_event(
                 CacheMiss(
-                    conn_name=self.nice_connection_name,
+                    conn_name=self.nice_connection_name(),
                     database=database,
                     schema=schema
                 )

--- a/core/dbt/events/base_types.py
+++ b/core/dbt/events/base_types.py
@@ -66,17 +66,6 @@ class ShowException():
         self.stack_info: Any = None
         self.extra: Any = None
 
-    def asdict(self):
-        d = dict()
-
-        for k, v in self.__dict__.items():
-            if isinstance(v, Exception):
-                d[k] = str(v)
-            else:
-                d[k] = v
-
-        return d
-
 
 # TODO add exhaustiveness checking for subclasses
 # can't use ABCs with @dataclass because of https://github.com/python/mypy/issues/5374
@@ -135,6 +124,17 @@ class Event(metaclass=ABCMeta):
     def get_invocation_id(cls) -> str:
         from dbt.events.functions import get_invocation_id
         return get_invocation_id()
+
+    # default dict factory for all events. can override on concrete classes.
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        d = dict()
+        for k, v in data:
+            if isinstance(v, Exception):
+                d[k] = str(v)
+            else:
+                d[k] = v
+        return d
 
 
 @dataclass  # type: ignore

--- a/core/dbt/events/base_types.py
+++ b/core/dbt/events/base_types.py
@@ -130,8 +130,12 @@ class Event(metaclass=ABCMeta):
     def asdict(cls, data: list) -> dict:
         d = dict()
         for k, v in data:
-            if isinstance(v, Exception):
+            # stringify all exceptions
+            if isinstance(v, Exception) or isinstance(v, BaseException):
                 d[k] = str(v)
+            # skip all binary data
+            elif isinstance(v, bytes):
+                continue
             else:
                 d[k] = v
         return d

--- a/core/dbt/events/base_types.py
+++ b/core/dbt/events/base_types.py
@@ -1,7 +1,6 @@
 from abc import ABCMeta, abstractmethod, abstractproperty
 from dataclasses import dataclass
 from datetime import datetime
-import json
 import os
 import threading
 from typing import Any, Optional
@@ -67,6 +66,17 @@ class ShowException():
         self.stack_info: Any = None
         self.extra: Any = None
 
+    def asdict(self):
+        d = dict()
+
+        for k, v in self.__dict__.items():
+            if isinstance(v, Exception):
+                d[k] = str(v)
+            else:
+                d[k] = v
+
+        return d
+
 
 # TODO add exhaustiveness checking for subclasses
 # can't use ABCs with @dataclass because of https://github.com/python/mypy/issues/5374
@@ -96,26 +106,6 @@ class Event(metaclass=ABCMeta):
     @abstractmethod
     def message(self) -> str:
         raise Exception("msg not implemented for Event")
-
-    # override this method to convert non-json serializable fields to json.
-    # for override examples, see existing concrete types.
-    #
-    # there is no type-level mechanism to have mypy enforce json serializability, so we just try
-    # to serialize and raise an exception at runtime when that fails. This safety mechanism
-    # only works if we have attempted to serialize every concrete event type in our tests.
-    def fields_to_json(self, field_value: Any) -> Any:
-        try:
-            json.dumps(field_value, sort_keys=True)
-            return field_value
-        except TypeError:
-            val_type = type(field_value).__name__
-            event_type = type(self).__name__
-            return Exception(
-                f"type {val_type} is not serializable to json."
-                f" First make sure that the call sites for {event_type} match the type hints"
-                f" and if they do, you can override Event::fields_to_json in {event_type} in"
-                " types.py to define your own serialization function to any valid json type"
-            )
 
     # exactly one time stamp per concrete event
     def get_ts(self) -> datetime:

--- a/core/dbt/events/functions.py
+++ b/core/dbt/events/functions.py
@@ -134,11 +134,13 @@ def event_to_serializable_dict(
     node_info = dict()
     log_line = dict()
     try:
-        log_line = dataclasses.asdict(e)
-        json.dumps(log_line)  # throwing away unprocessed result, just to make sure it works
+        if hasattr(e, 'asdict'):
+            log_line = dataclasses.asdict(e, dict_factory=type(e).asdict)  # type: ignore
+        else:
+            log_line = dataclasses.asdict(e)
     except AttributeError:
         event_type = type(e).__name__
-        raise Exception(
+        raise Exception(  # TODO this may hang async threads
             f"type {event_type} is not serializable to json."
             f" First make sure that the call sites for {event_type} match the type hints"
             f" and if they do, you can override the dataclass method `asdict` in {event_type} in"
@@ -151,7 +153,7 @@ def event_to_serializable_dict(
 
     for field, value in log_line.items():  # type: ignore[attr-defined]
         if field not in ["code", "report_node_data"]:
-            data[field] = e
+            data[field] = value
 
     event_dict = {
         'type': 'log_line',

--- a/core/dbt/events/functions.py
+++ b/core/dbt/events/functions.py
@@ -132,18 +132,26 @@ def event_to_serializable_dict(
 ) -> Dict[str, Any]:
     data = dict()
     node_info = dict()
-    if hasattr(e, '__dataclass_fields__'):
-        if isinstance(e, NodeInfo):
-            node_info = dataclasses.asdict(e.get_node_info())
+    log_line = dict()
+    try:
+        log_line = dataclasses.asdict(e)
+        json.dumps(log_line)  # throwing away unprocessed result, just to make sure it works
+    except AttributeError:
+        event_type = type(e).__name__
+        raise Exception(
+            f"type {event_type} is not serializable to json."
+            f" First make sure that the call sites for {event_type} match the type hints"
+            f" and if they do, you can override the dataclass method `asdict` in {event_type} in"
+            " types.py to define your own serialization function to a dictionary of valid json"
+            " types"
+        )
 
-        for field, value in dataclasses.asdict(e).items():  # type: ignore[attr-defined]
-            if field not in ["code", "report_node_data"]:
-                _json_value = e.fields_to_json(value)
+    if isinstance(e, NodeInfo):
+        node_info = dataclasses.asdict(e.get_node_info())
 
-                if not isinstance(_json_value, Exception):
-                    data[field] = _json_value
-                else:
-                    data[field] = f"JSON_SERIALIZE_FAILED: {type(value).__name__, 'NA'}"
+    for field, value in log_line.items():  # type: ignore[attr-defined]
+        if field not in ["code", "report_node_data"]:
+            data[field] = e
 
     event_dict = {
         'type': 'log_line',

--- a/core/dbt/events/functions.py
+++ b/core/dbt/events/functions.py
@@ -134,10 +134,7 @@ def event_to_serializable_dict(
     node_info = dict()
     log_line = dict()
     try:
-        if hasattr(e, 'asdict'):
-            log_line = dataclasses.asdict(e, dict_factory=type(e).asdict)  # type: ignore
-        else:
-            log_line = dataclasses.asdict(e)
+        log_line = dataclasses.asdict(e, dict_factory=type(e).asdict)
     except AttributeError:
         event_type = type(e).__name__
         raise Exception(  # TODO this may hang async threads

--- a/core/dbt/events/types.py
+++ b/core/dbt/events/types.py
@@ -878,7 +878,7 @@ class HandleInternalException(ShowException, DebugLevel, Cli, File):
 
 
 @dataclass
-class MessageHandleGenericException(ShowException, ErrorLevel, Cli, File):
+class MessageHandleGenericException(ErrorLevel, Cli, File):
     build_path: str
     unique_id: str
     exc: Exception
@@ -1277,7 +1277,7 @@ class InvalidRefInTestNode(WarnLevel, Cli, File):
 
 
 @dataclass
-class RunningOperationCaughtError(ShowException, ErrorLevel, Cli, File):
+class RunningOperationCaughtError(ErrorLevel, Cli, File):
     exc: Exception
     code: str = "Q001"
 
@@ -1286,7 +1286,7 @@ class RunningOperationCaughtError(ShowException, ErrorLevel, Cli, File):
 
 
 @dataclass
-class RunningOperationUncaughtError(ShowException, ErrorLevel, Cli, File):
+class RunningOperationUncaughtError(ErrorLevel, Cli, File):
     exc: Exception
     code: str = "FF01"
 
@@ -1303,7 +1303,7 @@ class DbtProjectError(ErrorLevel, Cli, File):
 
 
 @dataclass
-class DbtProjectErrorException(ShowException, ErrorLevel, Cli, File):
+class DbtProjectErrorException(ErrorLevel, Cli, File):
     exc: Exception
     code: str = "A010"
 
@@ -1320,7 +1320,7 @@ class DbtProfileError(ErrorLevel, Cli, File):
 
 
 @dataclass
-class DbtProfileErrorException(ShowException, ErrorLevel, Cli, File):
+class DbtProfileErrorException(ErrorLevel, Cli, File):
     exc: Exception
     code: str = "A012"
 
@@ -1376,7 +1376,7 @@ class CatchableExceptionOnRun(ShowException, DebugLevel, Cli, File):
 
 
 @dataclass
-class InternalExceptionOnRun(ShowException, DebugLevel, Cli, File):
+class InternalExceptionOnRun(DebugLevel, Cli, File):
     build_path: str
     exc: Exception
     code: str = "W003"
@@ -1406,7 +1406,7 @@ class PrintDebugStackTrace(ShowException, DebugLevel, Cli, File):
 
 
 @dataclass
-class GenericExceptionOnRun(ShowException, ErrorLevel, Cli, File):
+class GenericExceptionOnRun(ErrorLevel, Cli, File):
     build_path: str
     unique_id: str
     exc: Exception
@@ -2501,7 +2501,7 @@ class GeneralWarningMsg(WarnLevel, Cli, File):
 
 
 @dataclass
-class GeneralWarningException(ShowException, WarnLevel, Cli, File):
+class GeneralWarningException(WarnLevel, Cli, File):
     exc: Exception
     log_fmt: str
     code: str = "Z047"

--- a/core/dbt/events/types.py
+++ b/core/dbt/events/types.py
@@ -617,6 +617,7 @@ class SchemaCreation(DebugLevel, Cli, File):
     def asdict(cls, data: list) -> dict:
         return dict((k, str(v)) for k, v in data)
 
+
 @dataclass
 class SchemaDrop(DebugLevel, Cli, File):
     relation: BaseRelation
@@ -694,8 +695,9 @@ class DropCascade(DebugLevel, Cli, File, Cache):
             if isinstance(v, list):
                 d[k] = [str(x) for x in v]
             else:
-                d[k] = str(v)
+                d[k] = str(v)  # type: ignore
         return d
+
 
 @dataclass
 class DropRelation(DebugLevel, Cli, File, Cache):

--- a/core/dbt/events/types.py
+++ b/core/dbt/events/types.py
@@ -115,14 +115,6 @@ class MainEncounteredError(ErrorLevel, Cli):
     def message(self) -> str:
         return f"Encountered an error:\n{str(self.e)}"
 
-    # overriding default json serialization for this event
-    def fields_to_json(self, val: Any) -> Any:
-        # equality on BaseException is not good enough of a comparison here
-        if isinstance(val, BaseException):
-            return str(val)
-
-        return val
-
 
 @dataclass
 class MainStackTrace(DebugLevel, Cli):
@@ -150,12 +142,11 @@ class MainReportArgs(DebugLevel, Cli, File):
     def message(self):
         return f"running dbt with arguments {str(self.args)}"
 
-    # overriding default json serialization for this event
-    def fields_to_json(self, val: Any) -> Any:
-        if isinstance(val, argparse.Namespace):
-            return str(val)
-
-        return val
+    def asdict(self) -> dict:
+        return {
+            'code': self.code,
+            'args': str(self.args)
+        }
 
 
 @dataclass
@@ -353,13 +344,6 @@ class SystemCouldNotWrite(DebugLevel, Cli, File):
             f"Could not write to path {self.path}({len(self.path)} characters): "
             f"{self.reason}\nexception: {self.exc}"
         )
-
-    # overriding default json serialization for this event
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -667,12 +651,11 @@ class AddRelation(DebugLevel, Cli, File, Cache):
     def message(self) -> str:
         return f"Adding relation: {str(self.relation)}"
 
-    # overriding default json serialization for this event
-    def fields_to_json(self, val: Any) -> Any:
-        if isinstance(val, _CachedRelation):
-            return str(val)
-
-        return val
+    def asdict(self) -> dict:
+        return {
+            'code': self.code,
+            'relation': str(self.relation)
+        }
 
 
 @dataclass
@@ -782,11 +765,11 @@ class AdapterImportError(InfoLevel, Cli, File):
     def message(self) -> str:
         return f"Error importing adapter: {self.exc}"
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val())
-
-        return val
+    def asdict(self) -> dict:
+        return {
+            'code': self.code,
+            'exc': str(self.exc)
+        }
 
 
 @dataclass
@@ -842,12 +825,6 @@ class ProfileLoadError(ShowException, DebugLevel, Cli, File):
     def message(self) -> str:
         return f"Profile not loaded due to error: {self.exc}"
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 @dataclass
 class ProfileNotFound(InfoLevel, Cli, File):
@@ -885,12 +862,6 @@ class CatchRunException(ShowException, DebugLevel, Cli, File):
         )
         return error
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 # TODO: Remove? (appears to be uncalled)
 @dataclass
@@ -901,17 +872,11 @@ class HandleInternalException(ShowException, DebugLevel, Cli, File):
     def message(self) -> str:
         return str(self.exc)
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 # TODO: Remove? (appears to be uncalled)
 
 
 @dataclass
-class MessageHandleGenericException(ErrorLevel, Cli, File):
+class MessageHandleGenericException(ShowException, ErrorLevel, Cli, File):
     build_path: str
     unique_id: str
     exc: Exception
@@ -926,12 +891,6 @@ class MessageHandleGenericException(ErrorLevel, Cli, File):
             prefix=ui.red(prefix),
             error=str(self.exc).strip()
         )
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 # TODO: Remove? (appears to be uncalled)
 
@@ -1109,12 +1068,6 @@ class ParsedFileLoadFailed(ShowException, DebugLevel, Cli, File):
 
     def message(self) -> str:
         return f"Failed to load parsed file from disk at {self.path}: {self.exc}"
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -1322,33 +1275,21 @@ class InvalidRefInTestNode(WarnLevel, Cli, File):
 
 
 @dataclass
-class RunningOperationCaughtError(ErrorLevel, Cli, File):
+class RunningOperationCaughtError(ShowException, ErrorLevel, Cli, File):
     exc: Exception
     code: str = "Q001"
 
     def message(self) -> str:
         return f'Encountered an error while running operation: {self.exc}'
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 @dataclass
-class RunningOperationUncaughtError(ErrorLevel, Cli, File):
+class RunningOperationUncaughtError(ShowException, ErrorLevel, Cli, File):
     exc: Exception
     code: str = "FF01"
 
     def message(self) -> str:
         return f'Encountered an error while running operation: {self.exc}'
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -1360,18 +1301,12 @@ class DbtProjectError(ErrorLevel, Cli, File):
 
 
 @dataclass
-class DbtProjectErrorException(ErrorLevel, Cli, File):
+class DbtProjectErrorException(ShowException, ErrorLevel, Cli, File):
     exc: Exception
     code: str = "A010"
 
     def message(self) -> str:
         return f"  ERROR: {str(self.exc)}"
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -1383,18 +1318,12 @@ class DbtProfileError(ErrorLevel, Cli, File):
 
 
 @dataclass
-class DbtProfileErrorException(ErrorLevel, Cli, File):
+class DbtProfileErrorException(ShowException, ErrorLevel, Cli, File):
     exc: Exception
     code: str = "A012"
 
     def message(self) -> str:
         return f"  ERROR: {str(self.exc)}"
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -1443,15 +1372,9 @@ class CatchableExceptionOnRun(ShowException, DebugLevel, Cli, File):
     def message(self) -> str:
         return str(self.exc)
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 @dataclass
-class InternalExceptionOnRun(DebugLevel, Cli, File):
+class InternalExceptionOnRun(ShowException, DebugLevel, Cli, File):
     build_path: str
     exc: Exception
     code: str = "W003"
@@ -1469,12 +1392,6 @@ the error persists, open an issue at https://github.com/dbt-labs/dbt-core
             note=INTERNAL_ERROR_STRING
         )
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 # This prints the stack trace at the debug level while allowing just the nice exception message
 # at the error level - or whatever other level chosen.  Used in multiple places.
@@ -1487,7 +1404,7 @@ class PrintDebugStackTrace(ShowException, DebugLevel, Cli, File):
 
 
 @dataclass
-class GenericExceptionOnRun(ErrorLevel, Cli, File):
+class GenericExceptionOnRun(ShowException, ErrorLevel, Cli, File):
     build_path: str
     unique_id: str
     exc: Exception
@@ -1503,12 +1420,6 @@ class GenericExceptionOnRun(ErrorLevel, Cli, File):
             error=str(self.exc).strip()
         )
 
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
-
 
 @dataclass
 class NodeConnectionReleaseError(ShowException, DebugLevel, Cli, File):
@@ -1519,12 +1430,6 @@ class NodeConnectionReleaseError(ShowException, DebugLevel, Cli, File):
     def message(self) -> str:
         return ('Error releasing connection for node {}: {!s}'
                 .format(self.node_name, self.exc))
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -1845,12 +1750,6 @@ class SQlRunnerException(ShowException, DebugLevel, Cli, File):
 
     def message(self) -> str:
         return f"Got an exception: {self.exc}"
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass
@@ -2596,7 +2495,7 @@ class GeneralWarningMsg(WarnLevel, Cli, File):
 
 
 @dataclass
-class GeneralWarningException(WarnLevel, Cli, File):
+class GeneralWarningException(ShowException, WarnLevel, Cli, File):
     exc: Exception
     log_fmt: str
     code: str = "Z047"
@@ -2605,12 +2504,6 @@ class GeneralWarningException(WarnLevel, Cli, File):
         if self.log_fmt is not None:
             return self.log_fmt.format(str(self.exc))
         return str(self.exc)
-
-    def fields_to_json(self, val: Any) -> Any:
-        if val == self.exc:
-            return str(val)
-
-        return val
 
 
 @dataclass

--- a/core/dbt/events/types.py
+++ b/core/dbt/events/types.py
@@ -613,6 +613,9 @@ class SchemaCreation(DebugLevel, Cli, File):
     def message(self) -> str:
         return f'Creating schema "{self.relation}"'
 
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 @dataclass
 class SchemaDrop(DebugLevel, Cli, File):
@@ -621,6 +624,10 @@ class SchemaDrop(DebugLevel, Cli, File):
 
     def message(self) -> str:
         return f'Dropping schema "{self.relation}".'
+
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 
 # TODO pretty sure this is only ever called in dead code
@@ -680,6 +687,15 @@ class DropCascade(DebugLevel, Cli, File, Cache):
     def message(self) -> str:
         return f"drop {self.dropped} is cascading to {self.consequences}"
 
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        d = dict()
+        for k, v in data:
+            if isinstance(v, list):
+                d[k] = [str(x) for x in v]
+            else:
+                d[k] = str(v)
+        return d
 
 @dataclass
 class DropRelation(DebugLevel, Cli, File, Cache):

--- a/core/dbt/events/types.py
+++ b/core/dbt/events/types.py
@@ -142,11 +142,9 @@ class MainReportArgs(DebugLevel, Cli, File):
     def message(self):
         return f"running dbt with arguments {str(self.args)}"
 
-    def asdict(self) -> dict:
-        return {
-            'code': self.code,
-            'args': str(self.args)
-        }
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 
 @dataclass
@@ -526,7 +524,7 @@ class Rollback(DebugLevel, Cli, File):
 
 @dataclass
 class CacheMiss(DebugLevel, Cli, File):
-    conn_name: Any  # TODO mypy says this is `Callable[[], str]`??  ¯\_(ツ)_/¯
+    conn_name: str
     database: Optional[str]
     schema: str
     code: str = "E013"
@@ -547,6 +545,14 @@ class ListRelations(DebugLevel, Cli, File):
 
     def message(self) -> str:
         return f"with database={self.database}, schema={self.schema}, relations={self.relations}"
+
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        d = dict()
+        for k, v in data:
+            if type(v) == list:
+                d[k] = [str(x) for x in v]
+        return d
 
 
 @dataclass
@@ -651,11 +657,9 @@ class AddRelation(DebugLevel, Cli, File, Cache):
     def message(self) -> str:
         return f"Adding relation: {str(self.relation)}"
 
-    def asdict(self) -> dict:
-        return {
-            'code': self.code,
-            'relation': str(self.relation)
-        }
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 
 @dataclass
@@ -765,11 +769,9 @@ class AdapterImportError(InfoLevel, Cli, File):
     def message(self) -> str:
         return f"Error importing adapter: {self.exc}"
 
-    def asdict(self) -> dict:
-        return {
-            'code': self.code,
-            'exc': str(self.exc)
-        }
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 
 @dataclass
@@ -2220,6 +2222,10 @@ class NodeFinished(DebugLevel, Cli, File, NodeInfo):
 
     def message(self) -> str:
         return f"Finished running node {self.unique_id}"
+
+    @classmethod
+    def asdict(cls, data: list) -> dict:
+        return dict((k, str(v)) for k, v in data)
 
 
 @dataclass

--- a/test/unit/test_events.py
+++ b/test/unit/test_events.py
@@ -108,10 +108,6 @@ class TestEventBuffer(TestCase):
     #         EVENT_HISTORY.count(UnitTestInfo(msg='Test Event 1', code='T006')) == 0
     #     )
 
-def dump_callable():
-    return dict()
-    
-
 def MockNode():
     return ParsedModelNode(
         alias='model_one',
@@ -221,10 +217,10 @@ sample_values = [
         old_key=_ReferenceKey(database="", schema="", identifier=""),
         new_key=_ReferenceKey(database="", schema="", identifier="")
     ),
-    DumpBeforeAddGraph(dump_callable),
-    DumpAfterAddGraph(dump_callable),
-    DumpBeforeRenameSchema(dump_callable),
-    DumpAfterRenameSchema(dump_callable),
+    DumpBeforeAddGraph(dict()),
+    DumpAfterAddGraph(dict()),
+    DumpBeforeRenameSchema(dict()),
+    DumpAfterRenameSchema(dict()),
     AdapterImportError(ModuleNotFoundError()),
     PluginLoadError(),
     SystemReportReturnCode(returncode=0),
@@ -429,7 +425,7 @@ class TestEventJSONSerialization(TestCase):
 
         # if we have everything we need to test, try to serialize everything
         for event in sample_values:
-            d = event_to_serializable_dict(event, lambda dt: dt.isoformat(), lambda x: x.message())
+            d = event_to_serializable_dict(event, lambda _: event.get_ts_rfc3339(), lambda x: x.message())
             try:
                 json.dumps(d)
             except TypeError as e:

